modules/message-lane/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+		/// All stored (unprocessed) inbound messages.
+		InboundMessages: map hasher(blake2_128_concat) MessageKey => Option<T::Payload>;
+		/// Map of lane id => outbound lane data.
+		OutboundLanes: map hasher(blake2_128_concat) LaneId => OutboundLaneData;
+		/// All queued outbound messages.
+		OutboundMessages: map hasher(blake2_128_concat) MessageKey => Option<T::Payload>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call where origin: T::Origin {
+		/// Send message over lane.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn send_message(
+			origin,
+			lane_id: LaneId,
+			payload: T::Payload,
+		) {
+			let _ = ensure_signed(origin)?;
+			let mut lane = outbound_lane::<T, I>(lane_id);
+			lane.send_message(payload);
+			lane.prune_messages(T::MaxHeadersToPruneAtOnce::get());
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+	// =========================================================================================
+	// === Exposed mutables ====================================================================
+	// =========================================================================================
+
+	/// Receive new TRUSTED lane messages.
+	///
+	/// Trusted here means that the function itself doesn't check whether message has actually
+	/// been sent through the other end of the channel. We only check that we are receiving
+	/// and processing messages in order here.
+	///
+	/// Messages vector is required to be sorted by nonce within each lane. Otherise messages
+	/// will be rejected.
+	pub fn receive_messages(messages: Vec<Message<T::Payload>>) -> MessageNonce {
+		let mut correct_messages = 0;
+		let mut processor = T::OnMessageReceived::default();
+		for message in messages {
+			let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+			if lane.receive_message(message.key.nonce, message.payload, &mut processor) {
+				correct_messages += 1;
+			}
+		}
+
+		correct_messages
+	}
+
+	/// Process stored lane messages.
+	///
+	/// Stops processing either when all messages are processed, or when processor returns
+	/// MessageAction::Queue.
+	pub fn process_lane_messages(lane_id: &LaneId, processor: &mut impl OnMessageReceived<T::Payload>) {
+		inbound_lane::<T, I>(*lane_id).process_messages(processor);
+	}
+
+	/// Receive TRUSTED proof of message receival.
+	///
+	/// Trusted here means that the function itself doesn't check whether the bridged chain has
+	/// actually received these messages.
+	///
+	/// The caller may break the channel by providing `latest_received_nonce` that is larger
+	/// than actual one. Not-yet-sent messages may be pruned in this case.
+	pub fn confirm_receival(lane_id: &LaneId, latest_received_nonce: MessageNonce) {
+		let mut lane = outbound_lane::<T, I>(*lane_id);
+		lane.confirm_receival(latest_received_nonce);
+		lane.prune_messages(T::MaxHeadersToPruneAtOnce::get());
```

Comment:
I've added 3 events + another entry point - `fn confirm_processing(nonce)`. So we may now relay confirm that the message has been processed.

Reply:
I was actually thinking about notifying pallet that sent a particular message about it's delivery. Imagining a pallet, which serves as some sort of scheduler/queue/ratelimitter which sends highest-priority messages when some previous messages were relayed. However this can potentially be done in `on_intiialize` of the pallet, where it just inspects the storage to see if messages where confirmed.
If you feel the events are useful - go ahead, I haven't put much thought into it.